<a href="https://colab.research.google.com/github/Anastasiamay/Tinkoff/blob/main/task_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Задание Python, сгенерировать данные с помощью модуля random и ответить на вопросы
0) Для генерации данных в random необходимо установить seed = "АДВИ_2023", длина каждого списка N = 1000;

1) index колонка: создать список из N последовательных целых элементов от 1 до 1000 включительно, далее, с помощью специальной функции random перемешать его;

2) groups колонка: создать список из N элементов, состоящий из значений из списка ["группа_1", "группа_2", "группа_3", "группа_4", "группа_5"], необходимо использовать функцию из random которая возвращает список элементов длины k, выбранных из последовательности population с перестановкой элементов;

3) uniform колонка: создать список из N элементов, состоящий из целых значение из равномерного распределения [50, 100] ;

4) gauss_1, gauss_2, gauss_3 колонки: Создать 3 списка из N элементов каждый, заполненные значениями из нормальных распределений с параметрами:

- μ=0,σ=1 -> третья колонка gauss_1;
- μ=0,σ=144-> четвертая колонка gauss_2;
- μ=50,σ=81 -> пятая колонка gauss_3

5) Создать pandas DataFrame, где (1) будет индексом, (2)-(4) колонки, названия колонок ["groups", "uniform", "gauss_1", "gauss_2", "gauss_3"] соответственно, не забывая проверить, что типы колонок должны соответствовать заявленным типам данных;

6) Для колонок gauss_1 и gauss_2 заменить значения на None:

- Для gauss_1, если индекс у значение делится на 121 без остатка;
- Для gauss_2, если дробная часть больше 0.95

7) Теперь, для пропущенных значений gauss_1 и gauss_2:

- Для gauss_2 None заменить на среднее по колонке;
- Удалить строчки из таблицы, где gauss_1 == None.

In [ ]:
import random
import numpy as np
import pandas as pd
import math
# установим seed
random.seed("АДВИ_2023")

# зададим длину списка
N = 1000

# сгенерируем данные для колонок
index_col = list(range(1, N+1))
random.shuffle(index_col)

groups_col = random.choices(["группа_1", "группа_2", "группа_3", "группа_4", "группа_5"], k=N)

uniform_col = []
for i in range(N):
    uniform_col.append(random.randint(50, 100))                 
              

gauss_1_col = []
for i in range(1000): 
    gauss_1_col.append(random.gauss(0, 1)) 
    
gauss_2_col = []
for i in range(1000): 
    gauss_2_col.append(random.gauss(0, 12))  
    
gauss_3_col = []
for i in range(1000): 
    gauss_3_col.append(random.gauss(50, 9))     


# создадим DataFrame
df4 = pd.DataFrame({
    "index": index_col,
    "groups": groups_col,
    "uniform": uniform_col,
    "gauss_1": gauss_1_col,
    "gauss_2": gauss_2_col,
    "gauss_3": gauss_3_col
})

# заменим значения в колонках gauss_1 и gauss_2 на None
df4.loc[df4["index"] % 121 == 0, "gauss_1"] = None
df4.loc[((np.array(df4["gauss_2"]) - np.floor(np.array(df4["gauss_2"]))) > 0.95) & (df4["gauss_2"] > 0), "gauss_2"] = None

# Замена пропущенных значений
df4["gauss_2"].fillna(df4["gauss_2"].mean(), inplace=True)

# Удаление строк с пропущенными значениями gauss_1
df4 = df4.dropna(subset=["gauss_1"])

# Установка индекса
df4 = df4.set_index("index")
df4.describe(include='all')

groups     uniform     gauss_1     gauss_2     gauss_3
count        992  992.000000  992.000000  992.000000  992.000000
unique         5         NaN         NaN         NaN         NaN
top     группа_4         NaN         NaN         NaN         NaN
freq         222         NaN         NaN         NaN         NaN
mean         NaN   74.876008   -0.039448   -0.473799   49.444662
std          NaN   14.395724    1.052419   12.069964    9.021853
min          NaN   50.000000   -3.933254  -40.394911   19.682732
25%          NaN   62.000000   -0.745181   -8.334432   43.492984
50%          NaN   75.000000   -0.004941   -0.434114   49.654875
75%          NaN   88.000000    0.634956    7.139698   55.627863
max          NaN  100.000000    3.040902   34.218271   78.079961

Для колонки gauss_2 посчитать среднее и среднеквадратическое отклонение.

-0.47 12.07

In [ ]:
df4

groups  uniform   gauss_1    gauss_2    gauss_3
index                                                   
476    группа_1       95  3.010291   2.156346  57.455851
315    группа_5       77  0.190277  -9.934550  62.694177
490    группа_4       96 -0.282161   9.211071  54.688895
748    группа_3       56  1.127080 -12.391793  30.192353
471    группа_5       52 -1.238270  10.938748  52.682052
...         ...      ...       ...        ...        ...
260    группа_4       57 -1.484164  -5.911598  49.330196
146    группа_4       78 -0.379041  -3.132954  60.329900
972    группа_1       52 -0.857206   0.835715  51.926114
161    группа_1       62 -0.195475   3.496498  66.310250
587    группа_5       60  0.093496  -9.164475  60.897843

[992 rows x 5 columns]

Сколько всего получилось записей в таблице после выполнения всех пунктов из условия?

992

In [ ]:
df4['groups'].value_counts()

группа_4    222
группа_1    215
группа_2    194
группа_3    184
группа_5    177
Name: groups, dtype: int64

Какая группа встречается чаще всего и сколько раз?

группа_4 222

In [ ]:
df4['gauss'] = (df4['gauss_1']+df4['gauss_2'])/df4['gauss_3']
percentile_70 = df4["gauss"].quantile(0.7)
percentile_70

0.10641825158661447

Создать новую колонку gauss где значения будут получаться из (gauss_1 + gauss_2) / gauss_3

Чему будет равен 70% перцентиль?

0.11

In [ ]:
group_min_uniform = df4.groupby("groups")["gauss_3"].median().sort_values()
group_min_uniform

groups
группа_4    49.069511
группа_3    49.459261
группа_1    49.876708
группа_5    49.936910
группа_2    49.990529
Name: gauss_3, dtype: float64

У какой группы наибольшее значение медианы для gauss_3?

группа_2 49.99

In [ ]:
q = df4["uniform"].quantile(0.9)

count = df4.loc[df4["uniform"] >= q, "uniform"].count()
count

118

Сколько записей из колонки uniform имеют значение не ниже значения 90% перцентиля?

118

In [ ]:
group_min_uniform = df4.groupby("groups")["gauss_1"].min().sort_values()
group_min_uniform

groups
группа_4   -3.933254
группа_1   -3.628448
группа_2   -2.692705
группа_5   -2.639733
группа_3   -2.606773
Name: gauss_1, dtype: float64

In [ ]:
group_min_uniform = df4.groupby("groups")["gauss_2"].min().sort_values()
group_min_uniform

groups
группа_5   -40.394911
группа_2   -38.365537
группа_3   -36.187520
группа_4   -29.901064
группа_1   -22.109394
Name: gauss_2, dtype: float64

У какой группы наибольшее минимальное значение uniform?

Если такого нет, выбрать группу с наибольшим минимальным значением в gauss_1.

группа_3 -36.19

In [ ]:
group_5 = df4.loc[df4["groups"] == "группа_5"]
earliest_index = group_5.sort_index().head(1).index.values[0]
print(earliest_index)

6


 Для группа_5 вывести самый "ранний" (наименьший) индекс.

 6